### PR TITLE
Global roles v2 feature tests

### PR DESCRIPTION
**Pull request by Priyashetty17**
## Issue: 
https://github.com/rancher/qa-tasks/issues/930

## Problem: 
- Adding automated tests listed in the issue ^^ for **Global Roles v2** feature 
 
## Solution
- Added few helper functions to list, create, update, delete RBAC resources (global roles, global role binding, cluster role template binding etc) using kube api 
- Added the following automated tests - [rancher/qa-tasks#930](https://github.com/rancher/qa-tasks/issues/930)
  - Fresh install - Create custom roles and inherit the role[nested check]
  - Update the existing user and add the global role
  - Create a global role with multiple inheritedClusterRoles
  - Fresh Install Create downstream clusters after the global role is added to user should list the clusters for the user
  - Delete the inherited cluster role and verify all users should lose access and CRTBs should be deleted
  - Delete a downstream cluster and verify the bindings are deleted
  - Delete user and verify rbs and grbs are removed
  - Delete globalrolebinding and ensure the CRTBs are deleted for the user
- Added README
 
## Testing
- Jenkins Run

### CODE REVIEW

File: tests/v2/validation/rbac/globalrolesv2/globalroles_v2_test.go
Comment (anupama2501) [inline tests/v2/validation/rbac/globalrolesv2/globalroles_v2_test.go:82]: I'm not sure if this is apt. For example, I created a global role with the inheritedClusterRoles:
  - cluster-member
  - projects-view

I noticed that there were 2 crtbs created but 3 crbs created in the downstream cluster namely:

- crb-cuapmrktvo                                         ClusterRole/cluster-member                                         
 - crb-pbpiprkini                                         ClusterRole/view                                                   
- crb-tg3w66tern                                         ClusterRole/projects-view                                          


First CRB has the CRTB for cluster-member and the other 2 CRBs have the CRTBs referring to the same. In this scenario the test would fail. Verifying expectedCRB and expectedRB might be a tad difficult as they differ for individual roles and won't be the same always..

File: tests/v2/validation/rbac/globalrolesv2/globalroles_v2_test.go
Comment (anupama2501) [inline tests/v2/validation/rbac/globalrolesv2/globalroles_v2_test.go:89]: Similar to CRBs, the expectedRBCount may vary with different roles.

File: tests/v2/validation/rbac/globalrolesv2/globalroles_v2_test.go
Comment (anupama2501) [inline tests/v2/validation/rbac/globalrolesv2/globalroles_v2_test.go:117]:
```suggestion
		RoleTemplateIDs: []string{roleOwner},
```

File: tests/v2/validation/rbac/globalrolesv2/globalroles_v2_test.go
Comment (anupama2501) [inline tests/v2/validation/rbac/globalrolesv2/globalroles_v2_test.go:65]: Might be better if we separate out globalRoleCreation and validations. I noticed in one of the tests below, we are adding all the validations. Instead we can create 2 functions one for creation and one for validations of global roles v2. That way we do not have to re write the validation part in any of the tests.

File: tests/v2/validation/rbac/globalrolesv2/globalroles_v2_test.go
Comment (anupama2501) [inline tests/v2/validation/rbac/globalrolesv2/globalroles_v2_test.go:154]: As mentioned earlier, if we do not combine creation and validation of global roles, we can avoid re writing these validations here.

**Review by anupama2501 — COMMENTED** [review]
Overall looks good. have few comments and questions.

File: tests/v2/validation/rbac/globalrolesv2/globalroles_v2.go
Comment (Priyashetty17) [reply tests/v2/validation/rbac/globalrolesv2/globalroles_v2.go:79]: For tests involving few operations specifically delete operations (Delete user, cluster, global role) it takes a few seconds to reflect in the CRTBs. Hence the poll.

File: tests/framework/extensions/kubeapi/rbac/create.go
Comment (Priyashetty17) [reply tests/framework/extensions/kubeapi/rbac/create.go:74]: yes, makes sense to have it as a constant instead of having it as a parameter here. I will update CRTB functions as well since this comment applies to CRTBs as well.

File: tests/v2/validation/rbac/globalrolesv2/globalroles_v2.go
Comment (anupama2501) [inline tests/v2/validation/rbac/globalrolesv2/globalroles_v2.go:68]: Please remove the comment here and from the other places. We do not need to add comments for helper functions.

File: tests/v2/validation/rbac/globalrolesv2/globalroles_v2_test.go
Comment (Priyashetty17) [reply tests/v2/validation/rbac/globalrolesv2/globalroles_v2_test.go:82]: Based on our discussion, we will leave this code for CRBs and RBs as is.

File: tests/v2/validation/rbac/globalrolesv2/globalroles_v2_test.go
Comment (Priyashetty17) [reply tests/v2/validation/rbac/globalrolesv2/globalroles_v2_test.go:89]: Based on our discussion, we will leave this code for CRBs and RBs as is.

File: tests/v2/validation/rbac/globalrolesv2/globalroles_v2_test.go
Comment (Priyashetty17) [reply tests/v2/validation/rbac/globalrolesv2/globalroles_v2_test.go:48]: Separated global role creation and RBAC resource validation into separate functions

File: tests/framework/extensions/clusters/clusters.go
Comment (slickwarren) [inline tests/framework/extensions/clusters/clusters.go:1241]: I think we have a `local` string const.

File: tests/v2/validation/rbac/globalrolesv2/globalroles_v2.go
Comment (slickwarren) [inline tests/v2/validation/rbac/globalrolesv2/globalroles_v2.go:89]: nit; we should have global consts defined for this that you could use

File: tests/v2/validation/rbac/globalrolesv2/globalroles_v2.go
Comment (slickwarren) [inline tests/v2/validation/rbac/globalrolesv2/globalroles_v2.go:159]: we are trying to move away from creating clusters as part of tests that aren't exclusively testing provisioning. Is there a way this can be done using a user-provided cluster via the 'cluster-name' yaml variable?

File: tests/v2/validation/rbac/globalrolesv2/globalroles_v2.go
Comment (slickwarren) [inline tests/v2/validation/rbac/globalrolesv2/globalroles_v2.go:180]: what about k3s? is it unsupported? Same for imported clusters or hosted?

File: tests/v2/validation/rbac/globalrolesv2/globalroles_v2_test.go
Comment (slickwarren) [inline tests/v2/validation/rbac/globalrolesv2/globalroles_v2_test.go:47]: nit, since this is a helper function, can you remove test from the name? its a little confusing when reading through the code. Seems more appropriate to just call it validateRBACResources

File: tests/v2/validation/rbac/globalrolesv2/globalroles_v2_test.go
Comment (slickwarren) [inline tests/v2/validation/rbac/globalrolesv2/globalroles_v2_test.go:140]: is this doing something different than the `listDownstreamClusters` public function you created?

File: tests/v2/validation/rbac/globalrolesv2/globalroles_v2_test.go
Comment (slickwarren) [inline tests/v2/validation/rbac/globalrolesv2/globalroles_v2_test.go:341]: since you're creating users with iCRs can you update the test name to reflect that this creates and deletes resources/users?

**Review by slickwarren — COMMENTED** [review]
overall this looks great, just a few small comments. Nice work!

File: tests/v2/validation/rbac/globalrolesv2/globalroles_v2.go
Comment (Priyashetty17) [reply tests/v2/validation/rbac/globalrolesv2/globalroles_v2.go:180]: All that we are trying to validate in the Global Roles v2 tests is the ability of the user with cluster owner (inherited role) to create clusters. It doesn't matter what cluster type it is.

File: tests/v2/validation/rbac/globalrolesv2/globalroles_v2.go
Comment (igomez06) [inline tests/v2/validation/rbac/globalrolesv2/globalroles_v2.go:1]: New test files need tags. I'll post the documentation in the slack channel

File: tests/v2/validation/rbac/globalrolesv2/globalroles_v2.go
Comment (igomez06) [reply tests/v2/validation/rbac/globalrolesv2/globalroles_v2.go:159]: What we can also do is move this specific test case into the provisioning package. I guess what makes the most sense? Would it make sense to add to the table in our existing provisioning tests?

Comment (igomez06) [reply tests/v2/validation/rbac/globalrolesv2/globalroles_v2.go:159]: After thinking about this a bit more, Priya is even necessary to actually create a cluster? Wouldn't it make more sense to just check if it's allowed ie checking the schema? Doing something like this? 
https://github.com/rancher/rancher/blob/5b42ca50475398112661c80ac2b9070d5bacfb24/tests/framework/pkg/clientbase/ops.go#L190-L197

Comment (Priyashetty17) [reply tests/v2/validation/rbac/globalrolesv2/globalroles_v2.go:159]: Israel, the focus of these tests extends a little beyond just assessing the user's capability to create a cluster. We are also validating that when a new cluster is created by the user with the inherited cluster role, the respective CRTBs, CRBs and RBs are created as well. This is key in these tests. Additionally, we will be performing cluster-owner checks where we create projects/namespaces within the cluster. This is yet to be implemented (https://github.com/rancher/qa-tasks/issues/1038) after Anu's refactoring RBAC PR is merged. 
Is it a hard requirement to not have cluster creation in tests other than provisioning tests?

Comment (igomez06) [reply tests/v2/validation/rbac/globalrolesv2/globalroles_v2.go:159]: Ah I see, I mean we would like to, but I also understand that tests will cross over to other areas hmmm ... @slickwarren @markusewalker any thoughts? Priya makes a great point, but would like your guys' input as well since you are the provisioning team.

Comment (igomez06) [reply tests/v2/validation/rbac/globalrolesv2/globalroles_v2.go:159]: As an addition to the note above, would it make sense to make provisioning tests based on these roles, and validate that the CRTBs, CRBs and RBs are created?  Please let me know your thoughts.

Comment (Priyashetty17) [reply tests/v2/validation/rbac/globalrolesv2/globalroles_v2.go:159]: These tests are primarily centered on RBAC, ensuring the creation of specific RBAC resources. I'm hesitant about transferring them to the provisioning package. Moving them could lead to significant code duplication, especially given the numerous RBAC-specific private helper functions we already have in the rbac and globalrolesv2 packages.

File: tests/framework/extensions/clusters/clusters.go
Comment (markusewalker) [inline tests/framework/extensions/clusters/clusters.go:1237]: nit - can the error message be in a const variable instead?

File: tests/framework/extensions/kubeapi/rbac/list.go
Comment (markusewalker) [inline tests/framework/extensions/kubeapi/rbac/list.go:68]: Is this specific comment needed? IMO, the first comment should suffice.

File: tests/framework/extensions/kubeapi/rbac/list.go
Comment (markusewalker) [inline tests/framework/extensions/kubeapi/rbac/list.go:95]: Same as the above comment.

File: tests/framework/extensions/kubeapi/rbac/list.go
Comment (markusewalker) [inline tests/framework/extensions/kubeapi/rbac/list.go:122]: same as above.

File: tests/framework/extensions/kubeapi/rbac/list.go
Comment (markusewalker) [inline tests/framework/extensions/kubeapi/rbac/list.go:149]: same as above.

File: tests/framework/extensions/kubeapi/rbac/rbac.go
Comment (markusewalker) [inline tests/framework/extensions/kubeapi/rbac/rbac.go:47]: nit - just make this as one comment.

File: tests/framework/extensions/kubeapi/rbac/rbac.go
Comment (markusewalker) [inline tests/framework/extensions/kubeapi/rbac/rbac.go:55]: Same, just make as one comment.

File: tests/framework/extensions/kubeapi/rbac/rbac.go
Comment (markusewalker) [inline tests/framework/extensions/kubeapi/rbac/rbac.go:63]: same as above

File: tests/framework/extensions/kubeapi/rbac/rbac.go
Comment (markusewalker) [inline tests/framework/extensions/kubeapi/rbac/rbac.go:71]: same as above.

File: tests/v2/validation/rbac/globalrolesv2/README.md
Comment (markusewalker) [inline tests/v2/validation/rbac/globalrolesv2/README.md:12]: This seems to be in YAML but is defined as JSON, can you clarify?

File: tests/v2/validation/rbac/globalrolesv2/globalroles_v2.go
Comment (markusewalker) [reply tests/v2/validation/rbac/globalrolesv2/globalroles_v2.go:159]: Maybe we can have an adhoc meeting offline regarding this? Might be easier to sync up and figure out best scenario to work out.

Comment (igomez06) [reply tests/v2/validation/rbac/globalrolesv2/globalroles_v2.go:159]: Sounds good, that way we can come to a consensus. Priya can you set up a meeting?

Comment (Priyashetty17) [reply tests/v2/validation/rbac/globalrolesv2/globalroles_v2.go:159]: Sure, will do. Since Anu is unavailable this week and Israel next week, I can set up a meeting for 11/27. However, should we delay this PR until after our meeting, OR is it acceptable to proceed as it is currently and make changes based on the outcome of our discussion later?

Comment (igomez06) [reply tests/v2/validation/rbac/globalrolesv2/globalroles_v2.go:159]: I agree, we already have a test like this in rbac. This way we can come up with a consensus for these tests, the rbac test, and future tests. If anybody disagrees let me know

Comment (Priyashetty17) [reply tests/v2/validation/rbac/globalrolesv2/globalroles_v2.go:159]: Thanks Israel! For the time being, I will keep these tests involving new cluster provisioning as is. Based on the outcome of our meeting, we can later create a qa task to update the tests.

File: tests/v2/validation/rbac/globalrolesv2/README.md
Comment (Priyashetty17) [reply tests/v2/validation/rbac/globalrolesv2/README.md]: Copy paste error :) Fixed it.

File: tests/v2/validation/rbac/globalrolesv2/globalroles_v2_test.go
Comment (Priyashetty17) [reply tests/v2/validation/rbac/globalrolesv2/globalroles_v2_test.go]: Good catch! Modified the tests doing this to use the extension **listDownstreamClusters** instead

File: tests/v2/validation/rbac/globalrolesv2/globalroles_v2_test.go
Comment (igomez06) [inline tests/v2/validation/rbac/globalrolesv2/globalroles_v2_test.go:1]: Just to make sure Priya, this test will run every week?

Comment (Priyashetty17) [reply tests/v2/validation/rbac/globalrolesv2/globalroles_v2_test.go:1]: Sorry, I assumed extended referred to long duration tests. This can be run bi-weekly since it takes more than 15 minutes to run. I am going to change the tag to "stress".

Comment (igomez06) [reply tests/v2/validation/rbac/globalrolesv2/globalroles_v2_test.go]: Ok sounds good, just making sure you understand the differences

**Review by igomez06 — DISMISSED** [review]
Tags, and extensions look good. I'll defer to team 1 for reviews of the tests.

**Review by brudnak — DISMISSED** [review]
lgtm!

**Review by slickwarren — APPROVED** [review]

**Review by markusewalker — APPROVED** [review]

**Review by anupama2501 — APPROVED** [review]
lgtm

**Review by brudnak — APPROVED** [review]

**Review by anupama2501 — APPROVED** [review]

**Pull request merged**
